File: _skels/go-ddd-skel/internal/items/service.go
// Service-layer logic for the /api/items resource.
package items

import (
	"context"
	"fmt"
	"strings"

	"github.com/example/go-ddd-skel/internal/shared"
)

// Service coordinates items.Repository for HTTP routes. Exposed
// only via this layer so handlers stay free of *gorm.DB.
type Service struct {
	repo Repository
}

// NewService constructs an items Service.
func NewService(repo Repository) *Service {
	return &Service{repo: repo}
}

// List returns every item ordered by recency.
func (s *Service) List(ctx context.Context) ([]Item, error) {
	return s.repo.List(ctx)
}

// Get fetches a single item by id.
func (s *Service) Get(ctx context.Context, id uint) (Item, error) {
	return s.repo.Get(ctx, id)
}

// Create inserts a new item from the supplied DTO.
func (s *Service) Create(ctx context.Context, dto CreateItemDTO) (Item, error) {
	if strings.TrimSpace(dto.Name) == "" {
		return Item{}, fmt.Errorf("%w: item name cannot be empty", shared.ErrValidation)
	}
	item := Item{
		Name:        dto.Name,
		Description: dto.Description,
		IsCompleted: dto.IsCompleted,
		CategoryID:  dto.CategoryID,
	}
	if err := s.repo.Save(ctx, &item); err != nil {
		return Item{}, err
	}
	return item, nil
}

// Complete is the idempotent flip-to-done flow used by
// POST /api/items/{id}/complete.
func (s *Service) Complete(ctx context.Context, id uint) (Item, error) {
	item, err := s.repo.Get(ctx, id)
	if err != nil {
		return item, err
	}
	item.IsCompleted = true
	if err := s.repo.Save(ctx, &item); err != nil {
		return item, err
	}
	return item, nil
}

File: _skels/rust-actix-ddd-skel/src/items/service.rs
//! Service-layer logic for `/api/items`. Holds an `Arc<dyn
//! ItemRepository>` — never a `DatabaseConnection`.

use std::sync::Arc;

use crate::items::repository::{Item, ItemRepository, NewItem};
use crate::shared::DomainError;

/// Input shape for `create`. Separate from the entity so the route's
/// payload type and the service's input shape are decoupled.
#[derive(Debug, Clone)]
pub struct NewItemDTO {
    pub name: String,
    pub description: Option<String>,
    pub is_completed: bool,
    pub category_id: Option<i32>,
}

/// Coordinates `ItemRepository` for HTTP routes.
#[derive(Clone)]
pub struct ItemsService {
    repo: Arc<dyn ItemRepository>,
}

impl ItemsService {
    pub fn new(repo: Arc<dyn ItemRepository>) -> Self {
        Self { repo }
    }

    pub async fn list(&self) -> Result<Vec<Item>, DomainError> {
        self.repo.list().await
    }

    pub async fn get(&self, id: i32) -> Result<Item, DomainError> {
        self.repo.get(id).await
    }

    pub async fn create(&self, dto: NewItemDTO) -> Result<Item, DomainError> {
        if dto.name.trim().is_empty() {
            return Err(DomainError::Validation(
                "item name cannot be empty".to_string(),
            ));
        }
        self.repo
            .create(NewItem {
                name: dto.name,
                description: dto.description,
                is_completed: dto.is_completed,
                category_id: dto.category_id,
            })
            .await
    }

    pub async fn complete(&self, id: i32) -> Result<Item, DomainError> {
        self.repo.complete(id).await
    }
}
